src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs

```diff
@@ -511,49 +512,120 @@ private void AutoCompleteCluster(object parameter)
+                            targetNodeFromCluster = nodeFromCluster;
+                        }
+                        index++;
+                    });
+
+                    ClusterResultItem.Topology.Connections.ToList().ForEach(connection =>
+                    {
+                        // Connect the nodes
+                        var sourceNode = clusterMapping[connection.StartNode.NodeId];
+                        var targetNode = clusterMapping[connection.EndNode.NodeId];
+                        // The port index is 1- based (currently a hack and not expected from service)
+                        var sourcePort = sourceNode.OutPorts.FirstOrDefault(p => p.PortModel.Index == connection.StartNode.PortIndex - 1);
+                        var targetPort = targetNode.InPorts.FirstOrDefault(p => p.PortModel.Index == connection.EndNode.PortIndex - 1);
+                        var commands = new List<DynamoModel.ModelBasedRecordableCommand>
+                        {
+                            new DynamoModel.MakeConnectionCommand(sourceNode.Id.ToString(), connection.StartNode.PortIndex - 1, PortType.Output, DynamoModel.MakeConnectionCommand.Mode.Begin),
+                            new DynamoModel.MakeConnectionCommand(targetNode.Id.ToString(), connection.EndNode.PortIndex - 1, PortType.Input, DynamoModel.MakeConnectionCommand.Mode.End),
+                        };
+                        commands.ForEach(c =>
+                        {
+                            try
+                            {
+                                wsViewModel.DynamoViewModel.Model.ExecuteCommand(c);
+                            }
+                            catch (Exception) { }
+                        });
+                    });
+
+                    // Connect the cluster to the original node and port
+                    var finalCommands = new List<DynamoModel.ModelBasedRecordableCommand>
+                    {
+                        new DynamoModel.MakeConnectionCommand(node.Id.ToString(), 0, PortType.Output, DynamoModel.MakeConnectionCommand.Mode.Begin),
+                        new DynamoModel.MakeConnectionCommand(targetNodeFromCluster?.Id.ToString(), ClusterResultItem.EntryNodeInPort, PortType.Input, DynamoModel.MakeConnectionCommand.Mode.End),
+                    };
+                    finalCommands.ForEach(c =>
+                    {
+                        try
+                        {
+                            wsViewModel.DynamoViewModel.Model.ExecuteCommand(c);
+                        }
+                        catch (Exception) { }
+                    });
+
+                    // AutoLayout should be called after all nodes are connected
+                    foreach(var node in clusterMapping.Values)
+                    {
+                        wsViewModel.DynamoViewModel.Model.AddToSelection(node.NodeModel);
+                    }
+                    wsViewModel.Model.DoGraphAutoLayout(false, true, node.Id);
```

Comment:
Should we ensure that the cluster nodes are at predefined position/distance before AutoLayout ?
I see that if a connection fails the orphan node is put far away from the query node.
But having them stacked in proximity we ensure the ripple effect of AutoLayout is minimum.

Reply:
Also we will need some debounce for AutoLayout to allow WPF to update controls customization - but maybe we can do the autolayout as part of a different PR.

Reply:
Thank, this is the part I think we have some future improvement to do, @aparajit-pratap told me the Autolayout may not work well in a multi-node setup so I am still not too sure what would be the magic bullet here.